File: src/Leeson7/Homework7/Java/Book.java
package Leeson7.Homework7.Java;

public class Book {

    private String title;
    private int year;
    private Author author;

    public Book(String title, int year, Author author) {
        this.title = title;
        this.author = author;
        this.year = year;
    }

    //Вывод строки: Название, автор и год издания.
//    public void getAboutBook() {
//        System.out.println("Книга: " + title +
//                "\nАвтор: " + this.author +
//                "\nГод: " + year + ".");
//    }

    //Переопределение метода equals
    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if ((obj == null) || (this.getClass() != obj.getClass())) return false;

        /* Можно ли так объявить для сравнения свойств? Или это лишнее создание объекта? */
        //Book temp = (Book) obj;
        //return (this.title == temp.title && this.author == temp.author)?true:false;

        if (this.title.equals(((Book) obj).title) && author.equals(((Book) obj).author))
            return true;
        else return false;
    }

    @Override
    public int hashCode() {
        return this.title.hashCode() + this.year;
    }

    @Override
    public String toString() {
        return this.title + " " + this.year + " ";
    }
}

File: src/Leeson7/Homework7/Kotlin/Book.kt
package Leeson7.Homework7.Kotlin

class Book (val title : String, val author : String, private val year : Int ) {

    fun getInfo(){
        println("Книга '$title', автор $author, год $year")
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other !is Book) return false
        if (other === this) return true
        return this.title == other.title && this.author == other.author
    }

    override fun hashCode(): Int {
        return title.hashCode() + year
    }

    override fun toString(): String {
        return "Title: $title, author: $author, year: $year"
    }
}

